pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,229 @@
+// It is also called when the IP is upserted with a new identity
+// It removes the prefix from the prefixLengths map
+// It is called with the identityToIpMutex lock held
+func (s *FQDNDataServer) deleteFromIdentityToIPLocked(identity *ipcache.Identity, cidr net.IPNet, prefix netip.Prefix) error {
+	if identity == nil {
+		return fmt.Errorf("identity is nil")
+	}
+
+	if ips, ok := s.currentIdentityToIP[identity.ID]; ok {
+		newIPs := slices.DeleteFunc(ips, func(existing net.IPNet) bool {
+			if existing.String() == cidr.String() {
+				s.prefixLengths.Delete([]netip.Prefix{prefix})
+				return true
+			}
+			return false
+		})
+		if len(newIPs) == 0 {
+			delete(s.currentIdentityToIP, identity.ID)
+		} else {
+			s.currentIdentityToIP[identity.ID] = newIPs
+		}
+	}
+	return nil
+}
+
+// UpdatePolicyRules updates the current state of the DNS rules with the given policies and sends the current state of the DNS rules to the client
+// This method is called:
+// 1. when the DNS rules are updated during the endpoint regeneration, we store the state of the DNS rules with flag rulesUpdate as true
+// 2. when the client subscribes to DNS policies, we send the current state of the DNS rules to the client(flag rulesUpdate as false)
+// 3. when the IP identity cache changes, we update the current state of the identity to IP mapping and send the current state of the DNS rules to
+// the client(flag rulesUpdate as false)
+// Note: this method is left empty on purpose and will be updated with the actual implementation in the future PRs for the standalone DNS proxy
+func (s *FQDNDataServer) UpdatePolicyRules(policies map[identity.NumericIdentity]policy.SelectorPolicy, rulesUpdate bool) error {
+	return nil
+}
+
+// UpdateMappingRequest updates the FQDN mapping with the given data
+// SDP sends the fqdn mapping to cilium agent
+// Steps to update the mapping:
+// 1. Get the endpoint from the IP
+// 2. If the endpoint is not found, return an error
+// 3. If the IPs are not empty, update the cilium agent with the mapping
+// Note: this method is left empty on purpose and will be updated with the actual implementation in the future PRs for the standalone DNS proxy
+func (s *FQDNDataServer) UpdateMappingRequest(ctx context.Context, mappings *pb.FQDNMapping) (*pb.UpdateMappingResponse, error) {
+	return &pb.UpdateMappingResponse{
+		Response: pb.ResponseCode_RESPONSE_CODE_NO_ERROR,
+	}, nil
+}
+
+// ListenAndServe starts the Standalone DNS Proxy gRPC server on the given port
+func (s *FQDNDataServer) ListenAndServe(ctx context.Context, health cell.Health) error {
+	listenErrs := make(chan error)
+	go func() {
+		defer close(listenErrs)
+
+		address := fmt.Sprintf("localhost:%d", s.port)
+		s.log.Info("Starting Standalone DNS Proxy server on", logfields.Address, address)
+		lis, err := net.Listen("tcp", address)
+		if err != nil {
+			s.log.Error("Failed to listen", logfields.Error, err)
+			listenErrs <- err
```

Comment:
return after writing to the error chan; it's not like you can expect to serve successfully after failing to listen.